### PR TITLE
[camera] Zoom functionality for Android and iOS

### PR DESCRIPTION
**Pull request by mvanbeusekom**
## Description

Adds zoom support to the camera implementations for Android and iOS. It implements the interfaces defined #3312.

This PR is currently still in draft state, as it still depends on:
- #3302 General platform implementations being merged.
- #3312 Updated `camera_platform_interface` being published to pub.dev.

## Related Issues

- flutter/flutter#28694

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

**Comment by danielroek** [comment]
@googlebot I consent.

**Comment by BeMacized** [comment]
@googlebot I consent.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 1]
```diff
@@ -538,6 +514,57 @@ public void close() {
...
+    if (captureRequestBuilder != null) {
+      captureRequestBuilder.set(CaptureRequest.SCALER_CROP_REGION, zoomCropPreview);
+      cameraCaptureSession.setRepeatingRequest(captureRequestBuilder.build(), null, null);
+    }
+
+    result.success(null);
+  }
+
+  //Calculate zoom area according input image size
+  private static Rect getZoomRect(float zoom, int imgWidth, int imgHeight) {
```
Comment (bparrishMines) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:551]: This logic probably needs a test

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 2]
```diff
@@ -538,6 +514,57 @@ public void close() {
...
+    int cropWidth = (int) (imgWidth / zoom) + 2 * 64;
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: You were not missing anything at all. I made a mistake here and copied this formula from an earlier PR. I have looked up the correct way to implement zoom functionality and updated the logic.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 1 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: I have moved the zoom logic into its own class `CameraZoom` and make sure it is fully tested. To do so I had to bring in Robolectric (to be able to correctly use the `android.graphics.Rect` class in my unit tests). This also required an update of Gradle. I updated Gradle to stay in sync with the current version Flutter is using on stable (Gradle: 3.5.0, Gradle wrapper: 5.6.2).

**Review by bparrishMines — APPROVED** [review]
LGTM

**Pull request merged**